### PR TITLE
Fix PyTorch separate compilation

### PR DESCRIPTION
**Pull request by malfet**
Looks like there is a bug in CUDA device linker, but kernels that uses `thust::sort_by_key` can not be linked with other kernels
    Solve the problem by splitting 5 thrust-heavy .cu files into `__torch_cuda_sp` library which is statically linked into `torch_cuda`
    For default compilation workflow it should not make any difference.
    
    Test Plan: Compile with `-DCUDA_SEPARABLE_COMPILATION=YES` and observe library size difference: 310Mb before, 173Mb after if compiled for sm_75

### CODE REVIEW

**Comment by ezyang** [comment]
This is really nice sleuthing, but there is 100% chance that this is going to bitrot without extending CI coverage for it and making it more clear in the source code when people should put things in SP. Should we turn on separate compilation by default on our CUDA CI? Or does this have to be another conifg.

**Comment by orionr** [comment]
I think @ezyang's point about putting this in our CI is a good one. Can we enable it by default or is that a bad thing?

File: aten/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ set(ATen_CPU_TEST_SRCS)
 set(ATen_CPU_INCLUDE)
 set(ATen_THIRD_PARTY_INCLUDE)
 set(ATen_CUDA_SRCS)
+set(ATen_CUDA_SRCS_SP)
```
Comment (orionr) [inline aten/CMakeLists.txt]: Nit: Can we make this name more descriptive? ATen_CUDA_SRCS_W_SORT_BY_KEY or something?

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -701,11 +702,19 @@ IF (USE_ROCM)
 ELSEIF(USE_CUDA)
   set(CUDA_LINK_LIBRARIES_KEYWORD PRIVATE)
   cuda_add_library(torch_cuda ${Caffe2_GPU_SRCS})
+  cuda_add_library(__torch_cuda_sp ${Caffe2_GPU_SRCS_SP} STATIC)
+  add_dependencies(__torch_cuda_sp Caffe2_PROTO)
+  if (INTERN_BUILD_ATEN_OPS)
+  add_dependencies(__torch_cuda_sp aten_op_header_gen)
```
Comment (orionr) [inline caffe2/CMakeLists.txt]: Nit: spaces

**Review by orionr — COMMENTED** [review]
It's pretty amazing this buys us so much! Awesome.

**Comment by seemethere** [comment]
Yeah what are the downsides of enabling this by default?

File: caffe2/CMakeLists.txt (hunk 2 above)
Comment (malfet) [reply caffe2/CMakeLists.txt]: I understand the indent feedback, but what are the preferred formatting in `CMakeLists.txt`?
I.e.: `if (foo)` vs `if(foo)` vs `IF(foo)` vs `IF (foo)`. My personal preference is former (similar to C), but it doesn't seem that any particular style is followed in CMakeLists.txt

Comment (malfet) [reply caffe2/CMakeLists.txt]: Actually, pushed commit to the same branch that removes spaces

**Comment by malfet** [comment]
@seemethere, build times are slightly longer, also I'm not sure if it might have any negative effect on the GPU kernel performance (it shouldn't, but who knows)
I plan to enable it by default for some builds in a follow up PR

**Comment by seemethere** [comment]
I compiled this with CUDA 10.2 but it appears to be missing something

Compiled with:
```
time CMAKE_ARGS="-DCUDA_SEPARABLE_COMPILATION=YES" python setup.py bdist_wheel -d "${TMP_DIR}" | tee out.log
```

Error received after doing a `pip install *.whl`

```
❯ python
Python 3.8.1 (default, Jan  8 2020, 22:29:32)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/home/eliuriegas/anaconda3/envs/test38/lib/python3.8/site-packages/torch/__init__.py", line 126, in <module>
    _load_global_deps()
  File "/private/home/eliuriegas/anaconda3/envs/test38/lib/python3.8/site-packages/torch/__init__.py", line 84, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/private/home/eliuriegas/anaconda3/envs/test38/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcublasLt.so.10: cannot open shared object file: No such file or directory
```

**Comment by malfet** [comment]
@seemethere you need to make sure that CUDA libraries are available somewhere in dynamic linker search path

File: caffe2/CMakeLists.txt (hunk 2 above)
Comment (orionr) [reply caffe2/CMakeLists.txt]: I generally just match the file style, but let me see what you came up with.

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -702,7 +703,19 @@ if(USE_ROCM)
   endif()
 elseif(USE_CUDA)
   set(CUDA_LINK_LIBRARIES_KEYWORD PRIVATE)
-  cuda_add_library(torch_cuda ${Caffe2_GPU_SRCS})
+  if(CUDA_SEPARABLE_COMPILATION)
+    # Separate compilation fails when kernels using `thrust::sort_by_key`
```
Comment (orionr) [inline caffe2/CMakeLists.txt]: Nit: any external links to posts, issues, etc. you'd want to add here? Just might be something we'll all forget about in a years time or so.

**Review by orionr — APPROVED** [review]
This is great! Thank you. Just wondering if you want to add a little bit more documentation, but otherwise accepting to unblock.

File: caffe2/CMakeLists.txt (hunk 3 above)
Comment (malfet) [reply caffe2/CMakeLists.txt]: There are no external posts about this particular issue, but there are several issues reports against thrust/cub, for example https://github.com/thrust/thrust/issues/928

Comment (orionr) [reply caffe2/CMakeLists.txt]: Just throwing that link in there would be useful even though it's on a different repo.

**Comment by ezyang** [comment]
This broke ROCm builds:

```
00:59:17 CMakeFiles/Makefile2:3689: recipe for target 'caffe2/CMakeFiles/utility_ops_gpu_test.dir/all' failed
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaByteTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaHalfTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaLongTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaDoubleTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaIntTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaShortTensor_mode'
00:59:17 ../lib/libtorch_hip.so: undefined reference to `THCudaCharTensor_mode'
```

**Pull request closed without merging**